src/airflow/docker-compose.yaml

```diff
@@ -0,0 +1,232 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```

Comment:
Are we absolutely certain that we want to include the `docker-compose.yaml` as part of our code base? My issue with this is that this is an externally maintained file, with only a few custom changes introduced by us.

In the future, as Airflow continued to evolve, this config file might become obsolete in subtle ways, and in the worst case scenario cause all sorts of weird errors.

I'm still in favour of the approach where we fetch the fresh YAML file from Airflow, and then only introduce the changes we want. (This could even be automated using the `patch` command, so that the user doesn't have to do the changes manually.) In this case we can always be sure which changes are ours, and updates won't be a pain.

Reply:
I put quite a bit of thinking about this before implementing it and I checked online what the good practices are. My conclusions:

The main rationale is that you don't want every developer/user of the pipeline to repeat the same steps. In terms of overall benefit, we want to have as few steps as possible to spawn an airflow service.

Apparently, it's good practice to version the images/composer files, similar to why we version the `poetry.lock`. It keeps track of the version of the code/services that works and it can be tested. We could probably test more aspects of the airflow codebase but there are some tests already to ensure minimal functionality

My impression is that the alternative solution also has its own risks. Imagine a user "curling" the latest YAML and realising things don't run because the system has been changed in the latest YAML. Imagine this user under some pressure to get a run of the pipeline done. Also, we have introduced quite a few changes to the composer YAML (as anybody else is doing) which makes me think that the template YAML is just a very raw recipe that builds a very heavy general-purpose instance.

The cost is that we need to maintain the setup. One of the things, we added on top of the tests to ensure we keep things afloat is the dependabot rule that monitors the Dockerfile for version bumps. I don't think this covers everything, but it's a start and it will ensure PRs are opened and tested.

Reply:
Thank you for the explanation David. The points you make do make sense.

I wonder if we can have the best of two worlds by having the following set up:

* Store all custom modifications we make in a patch file, `docker-compose.yaml.patch`. This is the only file we directly make changes to.
* By curling a _specific_ version of the base YAML file and applying the patch, we obtainn the final file `docker-compose.yaml`. **This patched file is also stored in the repository,** and is the one used in all instructions.

So a regular user won't have to curl and patch anything, and the configuration is there; however, for us as developers, we can always keep track of all changes. And if a new version comes out, we can quickly update instead of performing a manual three way merge (old version — our modified version — new version)

Reply:
FYI **I don't think this is blocking** to merge this PR — in fact I'll go ahead and submit my approval. I'm just asking in general if you think this system makes sense, and if you do, I can implement it myself as a follow up PR

Reply:
it looks like a nice-to-have. It would improve the system. It's hard to know whether is needed without knowing how often there are breaking changes in Airflow composer yaml